Create PronounSeries and Gender models; delete Pronoun model

joshfeli opened the pull request:
# Create `PronounSeries` and `Gender` models; delete `Pronoun` model
## Overview of changes
This pull request features a number of changes outlined here:
- Removes the Pronoun model and all references to it (e.g., registry in `admin.py`, the `PronounTestCase` class in `tests.py`, the `PronounSerializer` class, etc)
- Adds the `PronounSeries` model
- Adds the `Gender` model
- Refactors the `DocumentManager` by placing it in its own file, `managers.py`
- Reconstructs a thorough sequence of migration files in the `migrations` module, replete with custom filenames and the ability to roll back to previous migrations

The commits show the existence of the `Name` and `NameSeries` models: while these ideas are saved in previous commits and migration files, we decided to remove these from the code after reevaluating the need for these models.

## Next steps
- *Add testing for the `PronounSeries` and `Gender` models.* While these models feature doctests, they could benefit from the creation of subclasses of `TestCase` for more thorough testing.
- *Resolve bugs in PyCharm configurations.* While working up to this PR, I had tried running `pytest` and creating a configuration to run `doctest` on the docstrings in `models.py`. In both scenarios, PyCharm did not recognize the relative file paths or even the `django` package. I recall running into an earlier problem with the `PronounSerializer`--perhaps this is worth some further debugging.

joshfeli committed:
Completed a sentence.

joshfeli committed:
Refactored LowercaseCharField to prevent potential aliasing error

joshfeli committed:
Added description to LowercaseCharField

Next steps (probably a new PR): writing tests for Pronoun and LowercaseCharField

joshfeli committed:
Setting up a test suite for Pronoun model; considering a __eq__ method

joshfeli committed:
Tweaked overridden description attribute in fields.py

joshfeli committed:
Resolved whitespace issue in models.py

joshfeli committed:
Setting up a test suite for LowercaseCharField

farooqashar committed:
updated Pronoun model from create_pronoun_model branch

farooqashar committed:
integrating the Pronoun model into PronounSeries Model

farooqashar committed:
work in progress pronoun series and gender models with many to many relationships

farooqashar committed:
added a potential way to integrate pronoun, pronoun series, and gender models

farooqashar committed:
fixed iteration over many to many fields of the pronoun series model and minor other bugs

farooqashar committed:
fixed minor bugs with the properties of the gender model

farooqashar committed:
implemented/fix bugs with the helper functions of the gender model

farooqashar committed:
added name and name series model in analogous to pronoun and pronoun series model

joshfeli committed:
Merge branch 'main' into create_gender_model

joshfeli committed:
Created a new migration file from merging main into create_gender_model

joshfeli committed:
Enforced a complete set of Pronoun instances in the PronounSeries model

Will probably add a helper method to display all pronoun identifiers as a set

joshfeli committed:
Revising the instance methods for the PronounSeries model

joshfeli committed:
Merge branch 'main' into create_gender_model

joshfeli committed:
Reformatter fields in PronounSeries model

joshfeli committed:
Trying to under a previous migrations mistake

Going back and attempting to reconstruct the migration files as they should have been while accounting for merges into main

joshfeli committed:
Got migrations to a stable state

For "Full-stack Squad":  with this commit, you will have to run "python manage.py migrate app zero" and then "python manage.py migrate" if I'm not mistaken. We can always go back and check the state of the DB using commands like "python manage.py showmigrations", but these commands worked for me.

joshfeli committed:
Removed all *references* to Name, NameSeries, and Pronoun models; ordered migrations up to that point

Anyone working on the create_gender_model branch up to this commit may have to delete their database if they started with a database that used previous versions of migration files.

Up to now, the Name, NameSeries, and Pronoun models are still present in models.py; the next step is deleting them safely.

joshfeli committed:
Removed last hidden reference to Pronoun and the Pronoun model; applied migration

joshfeli committed:
Removed Name and NameSeries models; applied migrations

joshfeli committed:
Refactored dunder methods for PronounSeries model

Refactored according to the changes with deleting the Pronoun model and restructuring data the PronounSeries model describes

joshfeli committed:
Modified doctests in Gender model

Considering setting `blank=True` and `null=True` for pronoun_series field to allow for subsequent adding of PronounSeries instances (since this is a ManyToMany relarion)

joshfeli committed:
Added PronounSeries and Gender models to registry

(in admin.py)

joshfeli committed:
Enforced fields in PronounSeries models to be filled; applied migration

Also added notes for next steps before a PR (or after one?)

joshfeli committed:
Removed PronounSerializer from serializers.py

Another sneaky reference to the deprecated Pronoun model!

joshfeli committed:
Merge branch 'main' into create_gender_model

joshfeli committed:
Moved DocumentManager to its own manager.py file

To help with code modularity

joshfeli committed:
Imported DocumentManager into models.py

joshfeli committed:
Removed PronounTestCase entirely; import Document model into tests.py

joshfeli committed:
Merged migration files from main and create_gender_model branches

MBJean commented on package.json:
```diff
@@ -39,7 +39,7 @@
     "file-loader": "^6.2.0",
     "mini-css-extract-plugin": "^1.6.0",
-    "webpack": "^5.38.1",
+    "webpack": "^5.39.0",
```
Is this version bump intentional, or just a side effect of some `npm` command run?

joshfeli replied:
I don't believe the version bump was intentional, but I do remember @farooqashar saying something about changing the `webpack` version to try and fix some bug..?

farooqashar replied:
Yes, there was an issue with the file `webpack-stats.json`not being generated after vising the browser and a workable solution was to downgrade the `webpack-bundle-tracker` package. I believe I did go back to the correct version for that package, but that automatically likely changed the webpack version as well. The version bump was not intentional, but was a side effect of downgrading and then upgrading the `webpack-bundle-tracker` package locally it seems like.

MBJean commented on backend/app/migrations/0001_initial.py:
```diff
@@ -17,13 +17,7 @@ class Migration(migrations.Migration):
             ],
         ),
-    ]
+    ]
```
This looks like a simple reformatting caused by some copying and pasting or something, but for the sake of setting the pattern, it's best if we don't modify old migrations. Would it be possible to back out this change? Or did it occur automatically for some other reason?

joshfeli replied:
I don't remember intentionally modifying that migration file--maybe it occurred automatically or when I was reconstructing the order of migrations. I can go back and revert this change.

MBJean commented on backend/app/migrations/0011_merge_document_gender_branches.py:
```diff
@@ -0,0 +1,14 @@
+    ]
+
+    operations = [
+    ]
```
Two considerations:

1. I'm not sure what this migration is doing, but it looks accidental.
2. Since the main branch doesn't contain any reference to the models that you created and then deleted in this branch, we don't necessarily need to include these migrations in the main branch. I think a more succinct approach would be to delete all of the migrations you've created here and rerun `makemigrations`. This will compile all changes in the `models.py` file into a single migration. If we do that, we'd want to squash merge your branch into main, so there'd only be a single commit and this new single migration file representing your work.

joshfeli replied:
That last migration file was created when I merged the `main` branch in before submitting the PR: the two files in the `dependencies` list are the ends of the two branches of files stemming from the `0002_document.py` migration. Django had thrown an error describing a merge conflict and prompted to run `py manage.py makemigrations --merge`, which created that file. After running `migrate`, everything seemed fine.

I don't really know what that migration file does (since the branches don't at all depend on each other), but it seemed to resolve that error.

With deleting the migrations file and running `makemigrations`, does that mean rolling back to `0002_document` and deleting all migrations after that, and then running `makemigrations`?

MBJean replied:
Going the deletion route would mean removing all migrations from this PR that aren't in the `main` branch (it looks like there's 4 atm), then rerunning `makemigrations`.

joshfeli replied:
Ok, I ended up squashing the branch of migration files into one that depended on `0002_document.py`. The `CommandError` I receive from the merge conflict with the branch of migrations concerning the `Document` model persists without the extra merge migration file `0005_merge_document_gender.py`. While that migration does not carry out any operations, I believe the ordering of migrations in the `dependencies` list is vital for resolving the order of execution for the two branches of migrations.

joshfeli committed:
Corrected syntax in Gender model doctests relating to ManyToMany relation to PronounSeries

joshfeli committed:
Restored the extra newline at the bottom of `0001_initial.py`

joshfeli committed:
Modified formatting in `0001.initial_py` to match copy in `main` branch

joshfeli committed:
Fixed formatting in `0001_initial.py`

joshfeli committed:
Squashed migration files created on create_gender_model branch

Some additional squashing may be required

joshfeli committed:
Deleted old branch of migration files

Django will now treat the squashed migration as a normal migration

joshfeli committed:
Completed replacement of old migrations; merged two branches

Had to run `python manage.py migrate --fake` to fake-apply the new migration files; the --fake tag should not be necessary if you've never run the old branches (though not tested) since you wouldn't have the PronounSeries and Gender tables in the DB yet. Another solution could be to delete the `db.sqlite3` file from the `backend` directory and migrate from scratch  (brute force method).

MBJean commented on backend/app/models.py:
```diff
@@ -8,40 +8,274 @@
+    """
+
+    label = models.CharField(max_length=60)
+    pronoun_series = models.ManyToManyField(PronounSeries)
```
On further reflection, is `ManyToMany` here correct? Is it possible for a single `Gender` to have multiple `PronounSeries`? I suppose that would allow us to do something like create a binary group gender that references multiple `PronounSeries`.

MBJean replied:
On reading through the commented tests below, I think that this makes sense. I think the commented-out methods below are worthwhile in this PR, then, and we should un-comment them. Otherwise, we have to go through some convolutions to get all of the relevant subject pronouns for a given `Gender`.

joshfeli replied:
Since the `PronounSeries` model only allows for 5 identifiers (versus 10), I thought of an example where a user might create a `Gender` instance with a set of singular pronouns and plural pronouns for some analysis.

MBJean reviewed (changes requested):
This looks great, thanks for your work! I left one more comment above of relevance: I think we should bring in the commented-out methods that allow us easy access to a Gender's PronounSeries's subjects.

Regarding your issues with testing: I recall that the `__init__.py` file in the `backend` directory was causing us problems before, and it looks like it's snuck back into this PR. Did you try getting rid of it and seeing if that allows you to build and run your tests accordingly?

joshfeli committed:
Changed `PronounSeries.get_all_pronouns()` method to the `all_pronouns` property; uncommented methods in `Gender` model

The doctests still have to be updated

joshfeli committed:
Deleted __init__.py file from `backend` directory

joshfeli committed:
Fixed bug in the `Gender.pronouns()` method

The method tried to perform a union with the *elements* of the desired set. The solution was the replace the inner for loop with a set union for each `series.all_pronouns` set.

MBJean reviewed (approved):
LGTM! Thanks for all your hard work, team, and for addressing my comments.

joshfeli committed:
Merge branch 'main' into create_gender_model
